External/FEXCore/include/FEXCore/Utils/ReentrantMutex.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <climits>
+#include <cstdint>
+#include <linux/futex.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+namespace FEXCore {
+  /**
+   * @brief A mutex class that can safely leave on signal and come back without hanging
+   */
+  class ReentrantMutex final {
```

Comment:
Is this a good name for this utility class? ReentrantMutex doesn't seem to deal with mutual exclusion, and the term seems to be overloaded already (https://en.wikipedia.org/wiki/Reentrant_mutex). My concurrency terminology isn't great either, but I think it's closer to a counting semaphore? (Or maybe RobustConditionVariable works, even though it's more of an integrated std::mutex+std::condition_variable+std::unique_lock)

Reply:
Naming is a hard thing. I wouldn't call it similar to a counting semaphore since there isn't any counting done, and I wouldn't call it robust since that conflates it with Linux's robust futexes.

Reply:
Names need not be perfect, but the current name is misleading in that it isn't used for mutual exclusion and it conflicts with established OS-agnostic terminology (vs a Linux-specific feature that only partly overlaps, i.e. only in the descriptor).

Calling it `<Something>+ConditionVariable` seems like the way to go IMO, since it's the same interface as `std::condition_variable` modulo minor differences (like the use of an internal `std::mutex` instead of a user-provided one).

How about `InterruptableCondVar`/`InterruptableConditionVariable`? It's a lengthy name (perhaps deservedly so for a specific use case) but it aptly describes what kind of object the class is representing *and* what distinguishes it from other synchronization primitives.

Reply:
Renamed class to `InterruptableConditionVariable`